src/cluster/src/check.rs

```diff
@@ -109,42 +125,249 @@ pub enum CheckError {
+    IncompatibleKubectlVersion {
+        /// The currently-installed helm version
+        installed: String,
+        /// The minimum required helm version
+        required: String,
+    },
+
+    /// Check permissions to create k8 resources
+    #[error("Permissions to create {resource} denied")]
+    PermissionError {
+        /// Name of the resource
+        resource: String,
+    },
+
+    /// Error while fetching create permissions for a resource
+    #[error("Unable to fetch permissions")]
+    FetchPermissionError,
+
+    /// One or more pre flight checks have failed
+    #[error("Some pre-install checks have failed.")]
+    PreFlightCheckError,
 }
 
-/// Getting server hostname from K8 context
-pub fn check_cluster_server_host() -> Result<(), CheckError> {
-    let config = K8Config::load()?;
-    let context = match config {
-        K8Config::Pod(_) => return Ok(()),
-        K8Config::KubeConfig(context) => context,
-    };
+/// Captures the status of the check
+pub enum StatusCheck {
+    /// Everything seems to be working as expected, check passed
+    Working(String),
+    /// Check failed due to an error, there is no work around
+    NotWorkingNoRemediation(CheckError),
+    /// Check failed due to an error, there is a work around fix
+    NotWorking(CheckError),
+}
 
-    let cluster_context = context
-        .config
-        .current_cluster()
-        .ok_or(CheckError::NoActiveKubernetesContext)?;
-    let server_url = cluster_context.cluster.server.to_owned();
-    let url =
-        Url::parse(&server_url).map_err(|source| CheckError::BadKubernetesServerUrl { source })?;
-    let host = url
-        .host()
-        .ok_or(CheckError::MissingKubernetesServerHost)?
-        .to_string();
-    if host.is_empty() {
-        return Err(CheckError::MissingKubernetesServerHost);
+#[async_trait]
+pub trait InstallCheck: Send + Sync + 'static {
+    /// perform check, if successful return success message, if fail, return fail message
+    async fn perform_check(&self) -> StatusCheck;
+}
+
+struct LoadableConfig;
+
+#[async_trait]
+impl InstallCheck for LoadableConfig {
+    async fn perform_check(&self) -> StatusCheck {
+        match check_cluster_server_host() {
+            Ok(_) => StatusCheck::Working(
+                "Kubernetes config is loadable and cluster hostname is not an IP address"
+                    .to_string(),
+            ),
+            Err(err) => StatusCheck::NotWorkingNoRemediation(err),
+        }
     }
-    if IpAddr::from_str(&host).is_ok() {
-        return Err(CheckError::KubernetesServerIsIp);
+}
+
+struct K8Version;
+
+#[async_trait]
+impl InstallCheck for K8Version {
+    async fn perform_check(&self) -> StatusCheck {
+        match k8_version_check() {
+            Ok(_) => StatusCheck::Working("Supported kubernetes version is installed".to_string()),
+            Err(err) => StatusCheck::NotWorkingNoRemediation(err),
```

Comment:
There is no remediation but there is info to user such as: `Please upgrade your Kubernetes clusters to at least...`